### PR TITLE
Implement Http/2 WebSockets

### LINKED ISSUE #7801: Kestrel support for WebSockets over HTTP/2
### Is your feature request related to a problem? Please describe.
I would like to be able to use WebSockets over HTTP2 - my understanding is that this is not currently supported, though I have not found any documentation about it or API support for it. My research shows that WebSockets and HTTP2 are independent within Kestrel and ASP.NET Core.

### Describe the solution you'd like
If `HttpProtocols.Http2` is enabled, and the WebSocketsMiddleware is enabled, support WebSockets over HTTP2 connections, per [RFC 8441](https://www.rfc-editor.org/rfc/rfc8441.txt).

### PR DESCRIPTION
**Pull request by Tratcher**
Fixes #7801

This adds support to Kestrel and the WebSocket middleware for HTTP/2 WebSockets. It's enabled by default in both layers and does not currently provide an opt-out. Browsers silently utilize HTTP/2 WebSockets when advertised by the server. HttpClient support is coming soon, when we have that we'll write some interop tests (#41895).

### CODE REVIEW

File: src/Middleware/WebSockets/src/WebSocketMiddleware.cs [hunk 1]
```diff
@@ -209,17 +239,22 @@ public static bool CheckSupportedWebSocketRequest(string method, IHeaderDictiona
...
 
+            if (!CheckWebSocketVersion(requestHeaders))
+            {
+                return false;
+            }
+
             var foundHeader = false;
 
-            var values = requestHeaders.GetCommaSeparatedValues(HeaderNames.SecWebSocketVersion);
+            var values = requestHeaders.GetCommaSeparatedValues(HeaderNames.Connection);
```
Comment (BrennanConroy) [inline src/Middleware/WebSockets/src/WebSocketMiddleware.cs:249]: https://datatracker.ietf.org/doc/html/rfc8441#section-5
> [RFC6455] requires the use of Connection and Upgrade header fields
   that are not part of HTTP/2.  They MUST NOT be included in the
   CONNECT request defined here.

File: src/Middleware/WebSockets/src/WebSocketMiddleware.cs [hunk 2]
```diff
@@ -249,28 +284,42 @@ public static bool CheckSupportedWebSocketRequest(string method, IHeaderDictiona
             {
                 return false;
             }
-            foundHeader = false;
 
-            values = requestHeaders.GetCommaSeparatedValues(HeaderNames.Upgrade);
+            return HandshakeHelpers.IsRequestKeyValid(requestHeaders.SecWebSocketKey.ToString());
```
Comment (BrennanConroy) [inline src/Middleware/WebSockets/src/WebSocketMiddleware.cs:288]: https://datatracker.ietf.org/doc/html/rfc8441#section-5
> Implementations using this extended CONNECT to bootstrap WebSockets
   do not do the processing of the Sec-WebSocket-Key and Sec-WebSocket-
   Accept header fields of [[RFC6455](https://datatracker.ietf.org/doc/html/rfc6455)] as that functionality has been
   superseded by the :protocol pseudo-header field.

We can avoid this work for H2

File: src/Middleware/WebSockets/src/HandshakeHelpers.cs [hunk 3]
```diff
@@ -25,12 +24,14 @@ internal static class HandshakeHelpers
...
-        headers.SecWebSocketAccept = CreateResponseKey(key);
+        if (isHttp1)
+        {
+            headers.Connection = HeaderNames.Upgrade;
+            headers.Upgrade = Constants.Headers.UpgradeWebSocket;
+            headers.SecWebSocketAccept = CreateResponseKey(key);
+        }
         if (!string.IsNullOrWhiteSpace(subProtocol))
         {
             headers.SecWebSocketProtocol = subProtocol;
```
Comment (BrennanConroy) [inline src/Middleware/WebSockets/src/HandshakeHelpers.cs:37]: Does something in the servers implementations make sure this header is lowercase?

File: src/Middleware/WebSockets/src/Microsoft.AspNetCore.WebSockets.csproj [hunk 4]
```diff
@@ -13,6 +13,7 @@
 
   <ItemGroup>
     <Reference Include="Microsoft.AspNetCore.Http.Extensions" />
+    <Reference Include="Microsoft.AspNetCore.Server.Kestrel.Core" />
```
Comment (BrennanConroy) [inline src/Middleware/WebSockets/src/Microsoft.AspNetCore.WebSockets.csproj:16]: 😢

File: src/Middleware/WebSockets/src/WebSocketMiddleware.cs [hunk 5]
```diff
@@ -159,7 +163,7 @@ public async Task<WebSocket> AcceptAsync(WebSocketAcceptContext acceptContext)
             }
 
             var key = _context.Request.Headers.SecWebSocketKey.ToString();
```
Comment (BrennanConroy) [inline src/Middleware/WebSockets/src/WebSocketMiddleware.cs:165]: We can avoid the `.ToString()` here for H2

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs [hunk 6]
```diff
@@ -233,18 +233,37 @@ private bool TryValidatePseudoHeaders()
...
         {
-            if (!String.IsNullOrEmpty(HttpRequestHeaders.HeaderScheme) || !String.IsNullOrEmpty(HttpRequestHeaders.HeaderPath))
+            // https://datatracker.ietf.org/doc/html/rfc8441#section-4
+            // HTTP/2 WebSockets
+            if (!StringValues.IsNullOrEmpty(HttpRequestHeaders.HeaderProtocol))
+            {
+                // On requests that contain the :protocol pseudo-header field, the :scheme and :path pseudo-header fields of the target URI MUST also be included.
+                if (StringValues.IsNullOrEmpty(HttpRequestHeaders.HeaderScheme) || StringValues.IsNullOrEmpty(HttpRequestHeaders.HeaderPath))
+                {
+                    ResetAndAbort(new ConnectionAbortedException("':protocol' requres ':scheme' and ':path'."), Http2ErrorCode.PROTOCOL_ERROR);
```
Comment (BrennanConroy) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs:245]:
```suggestion
                    ResetAndAbort(new ConnectionAbortedException("':protocol' requires ':scheme' and ':path'."), Http2ErrorCode.PROTOCOL_ERROR);
```

File: src/Middleware/WebSockets/src/HandshakeHelpers.cs (hunk 3 above)
Comment (Tratcher) [reply src/Middleware/WebSockets/src/HandshakeHelpers.cs:37]: The header name? Yeah, the HPack encoder does that.

File: src/Middleware/WebSockets/src/WebSocketMiddleware.cs (hunk 5 above)
Comment (Tratcher) [reply src/Middleware/WebSockets/src/WebSocketMiddleware.cs:165]: The ToString should be harmless in the common case. There should only be one value, and it will return that value without allocating.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.Generated.cs [hunk 7]
```diff
@@ -21,5 +21,6 @@ internal partial class Http2Connection
         private static ReadOnlySpan<byte> TeBytes => new byte[2] { (byte)'t', (byte)'e' };
         private static ReadOnlySpan<byte> TrailersBytes => new byte[8] { (byte)'t', (byte)'r', (byte)'a', (byte)'i', (byte)'l', (byte)'e', (byte)'r', (byte)'s' };
         private static ReadOnlySpan<byte> ConnectBytes => new byte[7] { (byte)'C', (byte)'O', (byte)'N', (byte)'N', (byte)'E', (byte)'C', (byte)'T' };
+        private static ReadOnlySpan<byte> ProtocolBytes => new byte[9] { (byte)':', (byte)'p', (byte)'r', (byte)'o', (byte)'t', (byte)'o', (byte)'c', (byte)'o', (byte)'l' };
```
Comment (Daniel-Genkin-MS-2) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.Generated.cs:24]: Why not do `private static ReadOnlySpan<byte> ProtocolBytes => ":protocol"u8;` instead of manually defining this? For context this is what we do in the HTTP/3 stuff.

Comment (Tratcher) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.Generated.cs:24]: The u8 syntax is brand new and this generated code hasn't been updated. That's a style change we'd do separately.

File: src/Middleware/WebSockets/src/WebSocketMiddleware.cs (hunk 1 above)
Comment (Tratcher) [reply src/Middleware/WebSockets/src/WebSocketMiddleware.cs:249]: CheckSupportedWebSocketRequest only covers HTTP/1.1, CheckSupportedWebSocketRequest covers HTTP/2.

File: src/Middleware/WebSockets/src/WebSocketMiddleware.cs (hunk 2 above)
Comment (Tratcher) [reply src/Middleware/WebSockets/src/WebSocketMiddleware.cs:288]: CheckSupportedWebSocketRequestH2 covers HTTP/2

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2WebSocketTests.cs [hunk 8]
```diff
@@ -1217,7 +1213,7 @@ public async Task ResponseWithHeadersTooLarge_AbortsConnection()
         _pair.Application.Output.Complete();
 
         await WaitForConnectionErrorAsync<HPackEncodingException>(ignoreNonGoAwayFrames: false, expectedLastStreamId: int.MaxValue, Http2ErrorCode.INTERNAL_ERROR,
-            SR.net_http_hpack_encode_failure);
+            SR.net_http_hpack_encode_failure); ;
```
Comment (Daniel-Genkin-MS-2) [inline src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2WebSocketTests.cs:1216]: I think there is a typo here, right?

File: src/Middleware/WebSockets/src/Microsoft.AspNetCore.WebSockets.csproj (hunk 4 above)
Comment (Tratcher) [reply src/Middleware/WebSockets/src/Microsoft.AspNetCore.WebSockets.csproj:16]: I think this is necessary to disable the request body rate limits. Since both of these assemblies are in the framework I don't think this causes any problems.

Comment (BrennanConroy) [reply src/Middleware/WebSockets/src/Microsoft.AspNetCore.WebSockets.csproj:16]: I know why it's there, but it means we don't have the correct abstractions for disabling limits in a non-server specific way.

Comment (Tratcher) [reply src/Middleware/WebSockets/src/Microsoft.AspNetCore.WebSockets.csproj:16]: Yes and no. No such feature is possible on the other servers so we never bothered abstracting it.

File: src/Http/Http.Features/src/IHttpRequestFeature.cs [hunk 9]
```diff
@@ -14,6 +14,16 @@ public interface IHttpRequestFeature
     /// </summary>
     string Protocol { get; set; }
 
+    /// <summary>
+    /// Gets or sets the :protocol field used with the CONNECT method.
+    /// See https://datatracker.ietf.org/doc/html/rfc8441.
+    /// </summary>
+    string? ConnectProtocol
```
Comment (davidfowl) [inline src/Http/Http.Features/src/IHttpRequestFeature.cs:21]: I want to go back in time and challenge why we made these all settable on the interface 😄

File: src/Middleware/WebSockets/src/Microsoft.AspNetCore.WebSockets.csproj (hunk 4 above)
Comment (davidfowl) [reply src/Middleware/WebSockets/src/Microsoft.AspNetCore.WebSockets.csproj:16]: Lets take the opportunity to do that now that we have a reason to.

Comment (halter73) [reply src/Middleware/WebSockets/src/Microsoft.AspNetCore.WebSockets.csproj:16]: Do we really have a reason to considering no other servers support this? Once there's at least one other server that does, I can see adding the abstraction.

Comment (davidfowl) [reply src/Middleware/WebSockets/src/Microsoft.AspNetCore.WebSockets.csproj:16]: Yes we do. I’m not with the dirtying the layering, approach. We don’t need to make an abstraction, but we need to move the feature interfaces out if kestrel. 

We should also not lose sight of our original design to build proper abstractions. HTTP.sys might end up supporting this.

Comment (halter73) [reply src/Middleware/WebSockets/src/Microsoft.AspNetCore.WebSockets.csproj:16]: It'd be more compelling if we still shipped Microsoft.AspNetCore.WebSockets or even Microsoft.AspNetCore.Http.Extensions as a package, but I supposed it does keep our project files a little cleaner.

Comment (halter73) [reply src/Middleware/WebSockets/src/Microsoft.AspNetCore.WebSockets.csproj:16]: Actually, why can't we implement `IHttpUpgradeFeature` for HTTP/2 and have the server automatically disable rate limits?

Comment (halter73) [reply src/Middleware/WebSockets/src/Microsoft.AspNetCore.WebSockets.csproj:16]: Okay, maybe not `IHttpUpgradeFeature` since that would be misleading but an entirely new feature that's similar. Maybe `IHttpConnectWithProtocolFeature` or something.

Comment (davidfowl) [reply src/Middleware/WebSockets/src/Microsoft.AspNetCore.WebSockets.csproj:16]: This is a discussion I was having with @Tratcher as well

Comment (Tratcher) [reply src/Middleware/WebSockets/src/Microsoft.AspNetCore.WebSockets.csproj:16]: > Yes we do. I’m not with the dirtying the layering, approach. We don’t need to make an abstraction, but we need to move the feature interfaces out if kestrel.

Why move them now if the namespace is still `Microsoft.AspNetCore.Server.Kestrel.Core.Features;`?

If another server is someday able to implement this, we can find the right abstraction then. I doubt the API would end up the same.

Comment (davidfowl) [reply src/Middleware/WebSockets/src/Microsoft.AspNetCore.WebSockets.csproj:16]: Until we merge the dlls into a single dll we should keep the layering. WebSockets shouldn't depend on Kestrel.

Comment (halter73) [reply src/Middleware/WebSockets/src/Microsoft.AspNetCore.WebSockets.csproj:16]: I think using `IHttpMinRequestBodyDataRateFeature` instead of something first-class like the `IHttpUpgradeFeature` used for HTTP/1.1 WebSockets is poor layering anyway. Other servers may support HTTP/2 WebSockets without supporting per-request rate limit control.

Comment (davidfowl) [reply src/Middleware/WebSockets/src/Microsoft.AspNetCore.WebSockets.csproj:16]: Perfect, even more reason to push this closer to the server.

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.FeatureCollection.cs [hunk 10]
```diff
@@ -284,6 +288,33 @@ async Task<Stream> IHttpUpgradeFeature.UpgradeAsync()
         return _bodyControl!.Upgrade();
     }
 
+    async Task<Stream> IHttpConnectFeature.AcceptAsync()
+    {
+        if (!IsConnectRequest)
+        {
+            // TODO Replace error messages and logs
```
Comment (Tratcher) [inline src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.FeatureCollection.cs:295]: Reminder, replace these log and exception messages.

File: src/Middleware/WebSockets/src/Microsoft.AspNetCore.WebSockets.csproj (hunk 4 above)
Comment (Tratcher) [reply src/Middleware/WebSockets/src/Microsoft.AspNetCore.WebSockets.csproj:16]: Ok, I've introduced a new IHttpConnectFeature for this scenario. How's it look?

File: src/Http/Http.Features/src/IHttpConnectFeature.cs [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    /// <summary>
+    /// The ':protocol' header included in the request.
+    /// </summary>
+    string? Protocol { get; }
+
+    /// <summary>
+    /// Send the response headers with a 200 status code and transition to opaque streaming.
+    /// </summary>
+    /// <returns>An opaque bidirectional stream.</returns>
+    Task<Stream> AcceptAsync();
```
Comment (davidfowl) [inline src/Http/Http.Features/src/IHttpConnectFeature.cs:26]:
```suggestion
    ValueTask<Stream> AcceptAsync();
```

New APIs should be `ValueTask`

File: src/Middleware/WebSockets/src/Microsoft.AspNetCore.WebSockets.csproj (hunk 4 above)
Comment (davidfowl) [reply src/Middleware/WebSockets/src/Microsoft.AspNetCore.WebSockets.csproj:16]: Much better, can ew delete this reference now?

File: src/Middleware/WebSockets/src/WebSocketMiddleware.cs [hunk 12]
```diff
@@ -245,28 +269,43 @@ public static bool CheckSupportedWebSocketRequest(string method, IHeaderDictiona
...
+        public static bool CheckSupportedWebSocketRequestH2(string method, string? protocol, IHeaderDictionary requestHeaders)
+        {
+            return HttpMethods.IsConnect(method)
+                && string.Equals(protocol, Constants.Headers.UpgradeWebSocket, StringComparison.OrdinalIgnoreCase)
+                && CheckWebSocketVersion(requestHeaders);
+        }
+
+        public static bool CheckWebSocketVersion(IHeaderDictionary requestHeaders)
+        {
+            var values = requestHeaders.GetCommaSeparatedValues(HeaderNames.SecWebSocketVersion);
```
Comment (davidfowl) [inline src/Middleware/WebSockets/src/WebSocketMiddleware.cs:295]: allocations?

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2WebSocketTests.cs [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+        // :authority = server.example.com
+        // sec-websocket-protocol = chat, superchat
+        // sec-websocket-extensions = permessage-deflate
+        // sec-websocket-version = 13
+        // origin = http://www.example.com
+        var headers = new[]
+        {
+            new KeyValuePair<string, string>(HeaderNames.Method, "CONNECT"),
+            new KeyValuePair<string, string>(HeaderNames.Protocol, "websocket"),
+            new KeyValuePair<string, string>(HeaderNames.Scheme, "http"),
```
Comment (greenEkatherine) [inline src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2WebSocketTests.cs:56]: Should be https - from comment?

Comment (Tratcher) [reply src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2WebSocketTests.cs:56]: Normally yes, but these in-memory tests bypass the TLS layer and Kestrel requires that the scheme match the transport.

File: src/Middleware/WebSockets/src/WebSocketMiddleware.cs (hunk 12 above)
Comment (Tratcher) [reply src/Middleware/WebSockets/src/WebSocketMiddleware.cs:294]: In the common case there will only be one value so the only allocation will be one string[1].

File: src/Middleware/WebSockets/src/HandshakeHelpers.cs [hunk 14]
```diff
@@ -17,15 +17,17 @@ internal static class HandshakeHelpers
...
+    public static void GenerateResponseHeaders(bool isHttp1, IHeaderDictionary requestHeaders, string? subProtocol, IHeaderDictionary responseHeaders)
     {
-        headers.Connection = HeaderNames.Upgrade;
-        headers.Upgrade = Constants.Headers.UpgradeWebSocket;
-        headers.SecWebSocketAccept = CreateResponseKey(key);
+        if (isHttp1)
+        {
+            responseHeaders.Connection = HeaderNames.Upgrade;
+            responseHeaders.Upgrade = Constants.Headers.UpgradeWebSocket;
+            responseHeaders.SecWebSocketAccept = CreateResponseKey(requestHeaders.SecWebSocketKey.ToString());
```
Comment (davidfowl) [inline src/Middleware/WebSockets/src/HandshakeHelpers.cs:26]: This isn't an extra allocation right?

File: src/Middleware/WebSockets/src/WebSocketMiddleware.cs (hunk 12 above)
Comment (davidfowl) [reply src/Middleware/WebSockets/src/WebSocketMiddleware.cs:294]: I'd like to see an allocation profile before and after?

File: src/Middleware/WebSockets/src/HandshakeHelpers.cs (hunk 14 above)
Comment (Tratcher) [reply src/Middleware/WebSockets/src/HandshakeHelpers.cs:26]: SecWebSocketKey.ToString()? No, just an explicit conversion from StringValues (singular) to string.

Comment (Tratcher) [reply src/Middleware/WebSockets/src/HandshakeHelpers.cs:26]: It won't, the lines you've commented on aren't even new, just re-arranged.

File: src/Middleware/WebSockets/samples/EchoApp/Startup.cs [hunk 15]
```diff
@@ -29,7 +29,7 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env, ILoggerF
         {
             if (context.WebSockets.IsWebSocketRequest)
             {
-                var webSocket = await context.WebSockets.AcceptWebSocketAsync();
+                var webSocket = await context.WebSockets.AcceptWebSocketAsync(new WebSocketAcceptContext() { DangerousEnableCompression = true });
```
Comment (davidfowl) [inline src/Middleware/WebSockets/samples/EchoApp/Startup.cs:32]: Why this change?

File: src/Middleware/WebSockets/test/UnitTests/Http2WebSocketTests.cs [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+
+        var testServer = host.GetTestServer();
+
+        var result = await testServer.SendAsync(httpContext =>
+        {
+            httpContext.Request.Method = HttpMethods.Connect;
+            httpContext.Features.Set<IHttpExtendedConnectFeature>(new ConnectFeature()
+            {
+                IsExtendedConnect = true,
+                Protocol = "WebSocket",
```
Comment (halter73) [inline src/Middleware/WebSockets/test/UnitTests/Http2WebSocketTests.cs:43]: Can we test sending and receiving data frames over the WebSocket and then gracefully closing it here? I understand that nothing should change relative to HTTP/1.1 WebSockets once you get the underlying Stream, but it's good to verify.

It's also tempting to open up the stream and connection windows in both directions Http2ConnectionTests and create a duplex Stream over the data frames for the request and wrap that in a client WebSocket so we can get better end-to-end testing of this scenario since we don't have any HttpClient or JS WebSocket based tests for the end-to-end. That'd be a more work, but probably not too much.

Comment (Tratcher) [reply src/Middleware/WebSockets/test/UnitTests/Http2WebSocketTests.cs:43]: I'm planning to add more functional tests once the HttpClient implementation is available. Without that it's pretty hard to write any realistic tests. The client team has asked us to merge this ASAP so they can use it develop the client.

**Review by halter73 — APPROVED** [review]
